### PR TITLE
identity: enforce key param and key existence on role creation

### PR DESCRIPTION
**Pull request by fairclothjm**
This fixes a bug found after merging: https://github.com/hashicorp/vault/pull/12151. 

A role should not be allowed to be created without a key parameter. Additionally, the key must exist.

### CODE REVIEW

**Review by pcman312 — COMMENTED** [review]
Looks good overall. Just a couple of small readability improvement suggestions

File: vault/identity_store_oidc_test.go [hunk 1]
```diff
@@ -18,6 +18,51 @@ import (
...
+	})
+	expectError(t, resp, err)
+	// validate error message
+	expectedStrings := map[string]interface{}{
+		"the key parameter is required": true,
+	}
+	expectStrings(t, []string{resp.Data["error"].(string)}, expectedStrings)
+}
+
+// TestOIDC_Path_OIDC_RoleNilKeyEntry tests taht a role cannot be created when
```
Comment (calvn) [inline vault/identity_store_oidc_test.go:42]:
```suggestion
// TestOIDC_Path_OIDC_RoleNilKeyEntry tests that a role cannot be created when
```

File: vault/identity_store_oidc_test.go [hunk 2]
```diff
@@ -18,6 +18,51 @@ import (
...
+	})
+	expectError(t, resp, err)
+	// validate error message
+	expectedStrings := map[string]interface{}{
+		"cannot find key \"test-key\"": true,
+	}
+	expectStrings(t, []string{resp.Data["error"].(string)}, expectedStrings)
+}
+
 // TestOIDC_Path_OIDCRoleRole tests CRUD operations for roles
```
Comment (calvn) [inline vault/identity_store_oidc_test.go:66]: Can we also add a test for updating a role with an empty key (which would not work) and one where the we update the role without providing a key value at all (which should work).

**Review by calvn — APPROVED** [review]
Suggestion for adding two additional test cases, implementation looks good though!

File: vault/identity_store_oidc_test.go (hunk 2 above)
Comment (fairclothjm) [reply vault/identity_store_oidc_test.go:66]: For the second test, we should get a failure right? Since `key` is a required param.

Comment (calvn) [reply vault/identity_store_oidc_test.go:66]: We should have decoded the key entry for updates, and because of `key, ok := d.GetOk("key")`, if it's `!ok` (missing) then the actual role.Key should be from the entry and non-empty.

File: changelog/12208.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+identity: handle creation of role without a key parameter
```
Comment (kalafut) [inline changelog/12208.txt:2]:
```suggestion
identity: disallow creation of role without a key parameter
```
Is that correct? I read "handle" as "support".

Comment (kalafut) [reply changelog/12208.txt:2]: I'm also thinking this should be more of a "changes" category. Folks could definitely be using the system like this today:

1. Make a role (no key)
2. Make a key (listing allowed_client_ids that needed to come from from step 1, so very reasonable to do this second)
3. Update the role with the key

This flow works today but will break with this change. We've said in our docs all along that the key was required, so we're really just meeting that contract so I think the change is OK, but it is an API break.

**Comment by kalafut** [comment]
We should mark the param as required in the path definition: https://github.com/hashicorp/vault/blob/a0e2449bd29bb60052d45e5487692cb0f57f2d6b/vault/identity_store_oidc.go#L243-L246

**Comment by calvn** [comment]
Since we're treating this as a breaking change, I milestoned it for 1.9

> This fixes a bug found after merging: #12151.

We found the odd behavior from #12151 but IIRC this existed before it (i.e. allowing key to be empty/missing). We can ship #12151 in 1.8.1 and hold this off for 1.9.

File: changelog/12208.txt (hunk 3 above)
Comment (fairclothjm) [reply changelog/12208.txt:2]: > Is that correct? I read "handle" as "support".

Yes, you are correct. I forgot to update the changelog after changing the implementation.

**Comment by fairclothjm** [comment]
@calvn 

> Since we're treating this as a breaking change, I milestoned it for 1.9
> 
> > This fixes a bug found after merging: #12151.
> 
> We found the odd behavior from #12151 but IIRC this existed before it (i.e. allowing key to be empty/missing). We can ship #12151 in 1.8.1 and hold this off for 1.9.

What I was trying to express there is that #12151 introduces a bug. An attempt to create a role with a key name that does not yet exist will fail with the error:

```
a role's token ttl cannot be longer than the verification_ttl of the key it references
```

This is what was discovered in the ENT tests that broke.

So if we are shipping #12151 in 1.8.1, I think there should be another fix to allow creating a role with a non-existent key.
